### PR TITLE
planner: add optimizer trace framework for physicalOptimize

### LINKED ISSUE #30334: add optimizer trace framework for physicalOptimize

### PR DESCRIPTION
**Pull request by rebelice**
### What problem does this PR solve?

Issue Number: close #30334

Problem Summary:

### What is changed and how it works?

Add optimize trace framework for physicalOptimize. In this pr, we add optimize trace common framework into physical optimize processing.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: util/tracing/opt_trace.go [hunk 1]
```diff
@@ -93,3 +93,34 @@ type CETraceRecord struct {
...
+	TP       string                       `json:"type"`
+	Children []*PhysicalOptimizeTraceInfo `json:"children"`
+	Property string                       `json:"property"`
+
+	// ExplainInfo should be implemented by each implemented LogicalPlan
+	ExplainInfo string `json:"info"`
+
+	BestTask   *TaskInfo   `json:"best"`
+	Candidates []*TaskInfo `json:"candidates"`
+}
```
Comment (Yisaer) [inline util/tracing/opt_trace.go:109]: Maybe we should maintain candidates like `map[string][]*TasnkInfo`, thus we can group by all the candidates by property?

File: util/tracing/opt_trace.go [hunk 2]
```diff
@@ -93,3 +93,34 @@ type CETraceRecord struct {
...
+
+type TaskInfo struct {
+	Cost float64 `json:"cost"`
+}
+
+// PhysicalOptimizeTracer indicates the trace for the whole physicalOptimize processing
+type PhysicalOptimizeTracer struct {
+	Root    *PhysicalOptimizeTraceInfo
+	Current *PhysicalOptimizeTraceInfo
+}
```
Comment (Yisaer) [inline util/tracing/opt_trace.go:119]: Add a comment that we maintain `PhysicalOptimizeTraceInfo` as a tree and let `Current` become `current`.

File: util/tracing/opt_trace.go [hunk 3]
```diff
@@ -93,3 +93,48 @@ type CETraceRecord struct {
...
+
+func (t *TaskInfo) SetCost(cost float64) {
+	t.Cost = cost
+}
+
+type PhysicalPlanList []string
+
+// PhysicalOptimizeTracer indicates the trace for the whole physicalOptimize processing
+type PhysicalOptimizeTracer struct {
+	Root    *PhysicalOptimizeTraceInfo
```
Comment (Yisaer) [inline util/tracing/opt_trace.go:130]: What is `Root` used for?

File: util/tracing/opt_trace.go [hunk 4]
```diff
@@ -93,3 +93,48 @@ type CETraceRecord struct {
...
+	Current *PhysicalOptimizeTraceInfo
+	Mapping map[string]PhysicalPlanList
```
Comment (Yisaer) [inline util/tracing/opt_trace.go:132]: I think `Current` is not necessary. We can directly maintain (LogicalPlan, Property)  -> Candidate Lists.

File: planner/core/find_best_task.go [hunk 5]
```diff
@@ -281,8 +289,54 @@ func (p *baseLogicalPlan) enumeratePhysicalPlans4Task(physicalPlans []PhysicalPl
...
+	if op.tracer == nil {
+		return
+	}
+	op.tracer.Current.BestTask = info
+}
+
+func (op *physicalOptimizeOp) appendCandidate(logicalPlan string, physicalPlan string) *tracing.TaskInfo {
+	if op.tracer == nil {
+		return nil
+	}
```
Comment (Yisaer) [inline planner/core/find_best_task.go:331]: I tend to pass `LogicalPlan` and PhsicalPlan here in order to skip running `fmt.Sprintf` if `op.tracer == nil`

**Review by Yisaer — COMMENTED** [review]
rest LGTM

File: planner/core/optimizer.go [hunk 6]
```diff
@@ -451,8 +451,19 @@ func physicalOptimize(logic LogicalPlan, planCounter *PlanCounterTp) (PhysicalPl
 		ExpectedCnt: math.MaxFloat64,
 	}
 
+	opt := defaultPhysicalOptimizeOption()
+	stmtCtx := logic.SCtx().GetSessionVars().StmtCtx
+	if stmtCtx.EnableOptimizeTrace {
+		tracer := &tracing.PhysicalOptimizeTracer{State: make(map[string]*tracing.PhysicalOptimizeTraceInfo), Mapping: make(map[string]tracing.PhysicalPlanList)}
+		opt = opt.withEnableOptimizeTracer(tracer)
+		opt.buildPhysicalOptimizeTraceInfo(logic, prop.String())
```
Comment (Yisaer) [inline planner/core/optimizer.go:459]: This seems not necessary.

File: util/tracing/opt_trace.go [hunk 7]
```diff
@@ -147,3 +149,44 @@ func DedupCETrace(records []*CETraceRecord) []*CETraceRecord {
...
+
+	// ExplainInfo should be implemented by each implemented LogicalPlan
+	ExplainInfo string `json:"info"`
+
+	BestTask   *TaskInfo   `json:"best"`
+	Candidates []*TaskInfo `json:"candidates"`
+}
+
+type TaskInfo struct {
+	Name     string                       `json:"name"`
```
Comment (Yisaer) [inline util/tracing/opt_trace.go:173]: Use `ID` and `Type` instead.

File: util/tracing/opt_trace.go [hunk 8]
```diff
@@ -147,3 +149,44 @@ func DedupCETrace(records []*CETraceRecord) []*CETraceRecord {
...
+func (t *TaskInfo) SetCost(cost float64) {
+	t.Cost = cost
+}
+
+type PhysicalPlanList []string
+
+// PhysicalOptimizeTracer indicates the trace for the whole physicalOptimize processing
+type PhysicalOptimizeTracer struct {
+	State   map[string]*PhysicalOptimizeTraceInfo
+	Mapping map[string]PhysicalPlanList
```
Comment (Yisaer) [inline util/tracing/opt_trace.go:187]: I think we can only maintain `State` by 

```golang
  // (logical plan) -> property -> physical plan candidates
  State   map[string]map[string]*PhysicalOptimizeTraceInfo
```

or 

```golang
  // (logical plan) -> (PhysicalOptimizeTraceInfo, property) list
  State   map[string][]*PhysicalOptimizeTraceInfo
```

File: planner/core/plan.go [hunk 9]
```diff
@@ -390,6 +390,11 @@ func (p *baseLogicalPlan) buildLogicalPlanTrace() *tracing.LogicalPlanTrace {
 	return planTrace
 }
 
+func buildPhysicalOptimizeTraceInfo(p Plan, prop string) *tracing.PhysicalOptimizeTraceInfo {
```
Comment (Yisaer) [inline planner/core/plan.go:393]:
```suggestion
func buildPhysicalOptimizeTraceInfo(p PhysicalPlan, prop string) *tracing.PhysicalOptimizeTraceInfo {
```

File: util/tracing/opt_trace.go (hunk 8 above)
Comment (Yisaer) [reply util/tracing/opt_trace.go:187]: And `Mapping` becomes not necessary.

File: planner/core/plan.go (hunk 9 above)
Comment (rebelice) [reply planner/core/plan.go]: It should be `LogicalPlan`, has been modified.

**Comment by rebelice** [comment]
@Yisaer updated all.

**Review by Yisaer — APPROVED** [review]
lgtm

File: planner/core/physical_plan_trace_test.go [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
```
Comment (hawkingrei) [inline planner/core/physical_plan_trace_test.go:17]: Please use testify to write new test case.

Comment (hawkingrei) [reply planner/core/physical_plan_trace_test.go:17]: You can see [this](https://github.com/pingcap/tidb/pull/31277)

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**